interpolation.go

```diff
@@ -60,3 +62,133 @@ func Interpolate(values []float64, size int) []float64 {
+ * https://www.npmjs.com/package/interp1?activeTab=code
+ *
+ * @param  xList        Array of independent sample points.
+ * @param  vList        Array of dependent values v(x) with length equal to xList.
+ * @param  qeuryList    Array of coordinates of the query points.
+ * @return              Interpolated values with length equal to queryList.
+ */
+func InterP1(xList, vList []float64, queryList []float64) ([]float64, error) {
+	if len(xList) != len(vList) {
+		return nil, fmt.Errorf("arrays of sample points xList and corresponding values vList have to have equal length.: %d vs. %d\n", len(xList), len(vList))
+	}
+	if len(xList) == 0 {
+		return nil, fmt.Errorf("arrays of sample points xList and corresponding values vList have to have length > 0.")
+	}
+
+	type Zip struct {
+		X, V float64
+	}
+	// Combine x and v
+	p := []Zip{}
+	for i := range xList {
+		p = append(p, Zip{xList[i], vList[i]})
+	}
+
+	// Sort asc
+	sort.Slice(p, func(i, j int) bool {
+		return p[i].X < p[j].X
+	})
+
+	// Check for double x values
+	for i, v := range p[1:] {
+		if p[i].X == v.X {
+			return nil, fmt.Errorf("two sample points have equal value %f. This is not allowed.", v.X)
+		}
+	}
+
+	// Split
+	sortedX, sortedV := []float64{}, []float64{}
+	for _, v := range p {
+		sortedX = append(sortedX, v.X)
+		sortedV = append(sortedV, v.V)
+	}
+
+	// Interpolate
+	r := []float64{}
+	for _, xq := range queryList {
+		// Determine index of range of query value.
+		index := binaryFindIndex(sortedX, xq)
+
+		// Check if value lies in interpolation range.
+		if index == -1.0 {
+			return nil, fmt.Errorf("query value %f lies outside of range. Extrapolation is not supported.", xq)
+		}
+
+		r = append(r, interpolateFloat(sortedV, index))
+	}
+
+	return r, nil
+}
+
+/**
+ * interpolateFloat applies linear interpolation on a value based on index.
+ * The index is allowed to be an non-integer. Used by InterP1D.
+ *
+ * This is ported from the js code which is ported from the matlab code:
+ * https://www.npmjs.com/package/interp1?activeTab=code
+ *
+ * @param  values Array of values to interpolate between.
+ * @param  index  Index of new to be interpolated value.
+ * @return        Interpolated value.
+ */
+func interpolateFloat(values []float64, index float64) float64 {
+	if index > float64(len(values)-1) {
+		return 0
+	}
+	prev := math.Floor(index)
+	next := math.Ceil(index)
+	lambda := index - prev
+	return (1-lambda)*values[int(prev)] + lambda*values[int(next)]
+}
+
+/**
+ * Finds the index of range in which a query value is included in a sorted
+ * array with binary search. Used by InterP1D.
+ *
+ * This is ported from the js code which is ported from the matlab code:
+ * https://www.npmjs.com/package/interp1?activeTab=code
+ *
+ * @param   values Array sorted in ascending order.
+ * @param   qValue Query value.
+ * @return         Index of range plus percentage to next index.
+ */
+func binaryFindIndex(values []float64, qValue float64) float64 {
```

Comment:
The naming and documentation of this method is confusing at best. It's just a binary search but which returns in "index" that isn't actually an integer index of the slice but a index + decimal value of how in-between the query value is of the value at index and the value at index+1.


It doesn't really matter but there is good opportunity to clarify this method to be less confusing. First thing is to drop the references to range. I'm not sure what range is being referred to at any point through out the function.

Reply:
Ya I totally agree I was confused by this name too, but I don't know what should be a good name, so I went with the js method name. Do you have any suggestions?

I think the `range` here just means `decimal value of how in-between the query value is of the value at index and the value at index+1`

Reply:
Eh, if it is basically just copied form js then I wouldn't bother making it clearer. If we really want to clean it up then lmk and I can come up with a better function name and documentation.

Reply:
k I would not want to spend time on this right now. If we want to improve we can do it in a new PR in the future.